### PR TITLE
Fixed y-range issue (#54), and added ggplot method.

### PR DESCRIPTION
**Pull request by Ax3man**
This PR contains a version of the `plot.keras_training_history` function. There is two main changes: there is now a `ggplot2` method, and the base method now takes validation metrics into account when calculating the y-range.

The code for the base plot has changed a bit, partly to fix the bug and partly to factor out common tasks between the two methods.

I have added `ggplot2` to Suggests.

## Plot examples

```
data(iris)

m <- keras_model_sequential() %>%
  layer_dense(12, input_shape = 4) %>%
  layer_dense(4, activation = 'sigmoid') %>%
  compile('adam', 'categorical_crossentropy', 'accuracy')

m2 <- m

h <- fit(m, as.matrix(iris[1:4]), to_categorical(iris[[5]]), 8, 100, validation_split = 0.2)

h2 <- fit(m2, as.matrix(iris[1:4]), to_categorical(iris[[5]]), 8, 50)
```

History `h` has validation data, `h2` does not.

Previously, the y range could get cause the validation to (partially) not be plotted, such as here:

`old_plot(h)`

old

This behavior is now fixed:

`plot(h, method = "base")`

base_new

The function now also checks whether `ggplot2` is available, and when it is, uses that package instead. 

For my system, these are equivalent and give the plot below:

```
plot(h)
plot(h, method = "auto")
plot(h, method = "ggplot2")
```

ggplot_val

When no validation data was used, the plots look like this:

`plot(h2, method = "base") # is exactly equivalent to the old behavior`

base_no_val

`plot(h2, method = "ggplot2")`

ggplot_no_val

### CODE REVIEW

**Comment by jjallaire** [comment]
Brilliant! A very, very nice improvement. I'll merge and then add you to `Authors@R` as a contributor. Thanks so much!

**Comment by Ax3man** [comment]
Excellent, glad to help! Feel free to hit me up if there are other plotting methods you'd like to add.

P.s. #54 can now be closed.

**Comment by jjallaire** [comment]
I noticed that for ggplot2 it's possible for the epoch axis to use non-integer labels, e.g. 

screen shot 2017-06-30 at 7 36 34 am

Is there a way to force it to use integer labels only?

**Comment by jjallaire** [comment]
The other cool thing to do would be an equivalent of the model visualization functions provided by the Python interface: https://keras.io/visualization/ which produces something like this:

multi-input-multi-output-graph

We could very easily provide an R interface to this but it has an underlying dependency that graphviz be installed and I think that's too much to ask. We could use DiagrammeR (http://rich-iannone.github.io/DiagrammeR/) which includes a JavaScript implementation of graphviz but that might be too heavyweight a dependency as well (we could always make it a Suggests as we did with ggplot2).

@gaborcsardi Do you know of any other legit options for doing these sorts of annotated graph visualizations? Maybe igraph is a good fit but the examples I've seen are not quite as text-oriented as the visualizations produced for models by the Python version of Keras.

**Comment by Ax3man** [comment]
Adding `ggplot2::scale_x_continuous(breaks = scales::pretty_breaks())` or even just `ggplot2::scale_x_continuous(breaks = pretty)` should help with the integer breaks.

I was thinking about diagrams actually. `DiagrammeR` indeed produces very pretty plots. I have been using it for my `phylopath` package. But I found control to be fairly limited, and vignettes become very large, and it is hard for users to save plots etc. So I ended up switching to `ggraph` instead.

There's also the `diagram` package, which is a bit older and uses base graphics.

It seems to be hard in R (including `igraph` and `ggraph`) to produce plots where the end of the arrows actually end up on the box, instead of ending at the centroid.

**Comment by gaborcsardi** [comment]
igraph is not really good for diagrams. Its (very limited) viz focuses more on actual (field) data and random graphs.

DiagrammeR has viz.js (https://github.com/mdaines/viz.js/) which is the same codebase as graphviz, so it is a better option. E.g. see the video at https://rich-iannone.github.io/DiagrammeR

**Comment by jjallaire** [comment]
To test out the `scale_x_continuous` I ran a training with 3 epochs and got these warnings:

```r
1: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  span too small.   fewer data values than degrees of freedom.
2: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  pseudoinverse used at 0.99
3: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  neighborhood radius 1.01
4: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  reciprocal condition number  0
5: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  There are other near singularities as well. 1.0201
6: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  span too small.   fewer data values than degrees of freedom.
7: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  pseudoinverse used at 0.99
8: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  neighborhood radius 1.01
9: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  reciprocal condition number  0
10: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  There are other near singularities as well. 1.0201
11: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  span too small.   fewer data values than degrees of freedom.
12: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  pseudoinverse used at 0.99
13: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  neighborhood radius 1.01
14: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  reciprocal condition number  0
15: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  There are other near singularities as well. 1.0201
16: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  span too small.   fewer data values than degrees of freedom.
17: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  pseudoinverse used at 0.99
18: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  neighborhood radius 1.01
19: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  reciprocal condition number  0
20: In simpleLoess(y, x, w, span, degree = degree, parametric = parametric,  ... :
  There are other near singularities as well. 1.0201
```

Perhaps we should only add the smoothing if there are a minimum number of epochs to plot?

`scale_x_continuous` works but still adds a decimal:

screen shot 2017-06-30 at 10 03 51 am
 
Would you mind putting together a patch to address both of these tweaks? (assuming there is a straightforward way to do it).

**Comment by Ax3man** [comment]
Sure, I can fix those soon. Perhaps a Boolean to control whether you want the smooth might be good.

Should I add documentation?

----- Reply message -----

**Comment by jjallaire** [comment]
Yes, a boolean to control smoothing for the ggplot2 method would be very useful.

I added some roxygen docs to the method already. Feel free to tweak/enhance as desired!
